src/lib/ppx_coda/register_event.ml

```diff
@@ -0,0 +1,223 @@
+open Core_kernel
+open Ppxlib
+
+let deriver = "register_event"
+
+module Foo = Result
+
+(* module in Structured_events library *)
+let module_name = "Structured_events"
+
+(* contains int_to_yojson, etc. *)
+let yojson_prims_modl = "Yojson_prims"
+
+let hash name =
+  let open Digestif.SHA256 in
+  let ctx0 = init () in
+  let ctx1 = feed_string ctx0 name in
+  get ctx1 |> to_raw_string
+
+(* get variable names for `{to,of}_yojson` functions, based on the type
+
+   the type must be an OCaml built-in, like `int` or `string`, or a type `t`
+    in some module, like `Foo.t`
+
+   missing: list or tuple types
+*)
+let make_yojson_name_of_core_type fname ty =
+  let add_lib_qualifier s = String.concat ~sep:"." [yojson_prims_modl; s] in
+  let fmt : ('a, Format.formatter, Base.unit, 'b) format4 =
+    "Field type must be given by an OCaml type constructor or tuple"
+  in
+  match ty.ptyp_desc with
+  | Ptyp_constr (lident, []) -> (
+    match lident.txt with
+    | Lident id ->
+        add_lib_qualifier (id ^ "_" ^ fname)
+    | Ldot (prefix, "t") ->
+        String.concat ~sep:"." (Longident.flatten_exn prefix @ [fname])
+    | _ ->
+        Location.raise_errorf ~loc:ty.ptyp_loc fmt )
+  | _ ->
+      Location.raise_errorf ~loc:ty.ptyp_loc fmt
+
+let to_yojson_of_core_type = make_yojson_name_of_core_type "to_yojson"
+
+let of_yojson_of_core_type = make_yojson_name_of_core_type "of_yojson"
+
+let core_kernel_result f = String.concat ["Core_kernel"; "Result"; f] ~sep:"."
+
+let type_ext_str ~options ~path:_ (ty_ext : type_extension) =
+  let add_module_qualifier s = String.concat ~sep:"." [module_name; s] in
+  let ctor, record, label_decls =
+    match ty_ext.ptyext_constructors with
+    | [ { pext_name
+        ; pext_kind= Pext_decl ((Pcstr_record labels as record), None)
+        ; _ } ] ->
+        (pext_name.txt, record, labels)
+    | _ ->
+        Location.raise_errorf ~loc:ty_ext.ptyext_path.loc
+          "Constructor in type extension must take a single record argument"
```

Comment:
I think we might want to allow no arguments here, even if we aren't handling the anonymous argument/tuple.